src/protocol/mod.rs

```diff
@@ -31,33 +34,39 @@ impl Step {
     pub fn new(command_matcher: CommandMatcher) -> Step {
         Step {
             command_matcher,
+            marker: None,
             stdout: vec![],
             exitcode: 0,
         }
     }
 
+    pub fn marker(&self) -> Option<Marker> {
+        let Step { marker, .. } = self;
+        *marker
+    }
+
     fn from_string(string: &str) -> R<Step> {
         Ok(Step::new(CommandMatcher::ExactMatch(Command::new(string)?)))
     }
 
-    fn add_exitcode(&mut self, object: &Hash) -> R<()> {
+    fn add_exitcode(&mut self, object: &HashNode) -> R<()> {
         if let Ok(exitcode) = object.expect_field("exitcode") {
             self.exitcode = exitcode.expect_integer()?;
         }
         Ok(())
     }
 
-    fn add_stdout(&mut self, object: &Hash) -> R<()> {
+    fn add_stdout(&mut self, object: &HashNode) -> R<()> {
         if let Ok(stdout) = object.expect_field("stdout") {
             self.stdout = stdout.expect_str()?.as_bytes().to_vec();
         }
         Ok(())
     }
 
-    fn parse(yaml: &Yaml) -> R<Step> {
-        match yaml {
-            Yaml::String(string) => Step::from_string(string),
-            Yaml::Hash(object) => {
+    fn parse(Node(yaml, marker): &Node) -> R<Step> {
```

Comment:
ooh, i didn't know you could destructure in function signatures

Reply:
Yep! But there is a weird exception where (AFAIK) you cannot destructure an argument if it is `self` because there is just no syntax for that.